notation.go

```diff
@@ -63,31 +42,188 @@ type SignOptions struct {
+
 	// SignatureMediaType is the envelope type of the signature.
-	// Currently both `application/jose+json` and `application/cose` are supported.
+	// Currently both `application/jose+json` and `application/cose` are
+	// supported.
 	SignatureMediaType string
+
+	// PluginConfig is a map of plugin configs.
+	PluginConfig map[string]string
 }
 
-// Validate does basic validation on VerifyOptions.
-func (opts VerifyOptions) Validate() error {
-	return nil
+// ValidationResult encapsulates the verification result (passed or failed)
+// for a verification type, including the desired verification action as
+// specified in the trust policy
+type ValidationResult struct {
+	// Type of verification that is performed
+	Type trustpolicy.ValidationType
+
+	// Action is the intended action for the given verification type as defined
+	// in the trust policy
+	Action trustpolicy.ValidationAction
+
+	// Error is set if there are any errors during the verification process
+	Error error
+}
+
+// VerificationOutcome encapsulates a signature blob's descriptor, its content,
+// the verification level and results for each verification type that was
+// performed.
+type VerificationOutcome struct {
+	// SignatureBlobDescriptor is descriptor of the signature envelope blob
+	SignatureBlobDescriptor *ocispec.Descriptor
+
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*ValidationResult
+
+	// Error that caused the verification to fail (if it fails)
+	Error error
 }
 
 // Verifier is a generic interface for verifying an artifact.
 type Verifier interface {
-	// Verify verifies the signature and returns the verified descriptor and
-	// metadata of the signed artifact.
-	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (Descriptor, error)
+	// Verify verifies the signature blob and returns the artifact
+	// descriptor upon successful verification.
+	// If signature == nil and the verification level is not 'skip', an error
+	// will be returned.
+	Verify(ctx context.Context, signature []byte, opts VerifyOptions) (ocispec.Descriptor, *VerificationOutcome, error)
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
```

Comment:
@yizha1 @patrickzheng200 @priteshbandi @rgnote I have security concerns on
```golang
func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
```
returning `[]*VerificationOutcome`.

What if the remote registry is malicious and returns millions of invalid signature manifests? It will create millions of objects in `[]*VerificationOutcome`. Eventually, the process may be out of memory and get killed. It affects the availability if notation-go is integrated with a server application.

Reply:
Several options here.
1. Remove `[]*VerificationOutcome` from the list of return parameters.
3. We set a limit on the number of attempted signatures.

Reply:
I think the underlying concern is: What's the maximum number of signatures we will evaluate (assuming all failed verification) before declaring overall verification as failed?
For the short term, we can put some reasonable limit(TBD - 50?) on the maximum number of signatures that the notation will evaluate. If an image contains more than that, the notation will fail verification. The catch here is that the order of signature artifact manifest descriptors returned by refers API should be deterministic.

Reply:
Added a limit number 50 for now.

Reply:
@priteshbandi per your suggestion, could you create an issue to track the configuration of "maximum number of signatures"?

Reply:
> Keep in mind that a specially crafted single signature may cause resource exhaustion too, for example, a signature with thousands of certificates in it.

That's covered since we have a size limit for a signature blob. Also, we have a size limit for signature manifest. Again, we also have a size limit for the return value of the referrers API as the underlying `oras-go` takes care of those security checks.

Reply:
had a discussion with @gokarnm on this and he suggested using timeout as the restricting mechanism which sounds more appt here.

We should also modify the `VerificationOutcome` behavior to return only successful signature verification outcome.

cc: @shizhMSFT @rgnote

Reply:
@priteshbandi Can you create an issue for discussions and tracking?

Reply:
I see it is tracked in #201